src/collapse/common.rs

```diff
@@ -0,0 +1,780 @@
+                    // so there's no need to check for a `SendError` here.
+                    let _ = tx_input.send(Some(buf));
+                    break;
+                }
+                let line = &buf[index..index + n];
+                index += n;
+                if self.is_end_of_stack(line) {
+                    // If we've reached the end of a stack, count it.
+                    nstacks += 1;
+                    if nstacks == nstacks_per_job {
+                        // If we've accumulated enough stacks to make up a chunk to send to the
+                        // worker threads, try to send it.
+                        let buf_capacity = usize::next_power_of_two(buf.capacity());
+                        let chunk = mem::replace(&mut buf, Vec::with_capacity(buf_capacity));
+                        if tx_input.send(Some(chunk)).is_err() {
+                            // If sending the chunk produces a `SendError`, this means that one
+                            // of the worker threads has errored, sent a signal to all the other
+                            // worker threads to shut down, and they have all shutdown, in which
+                            // case we know there will be an error waiting for us on the error
+                            // channel; so we should stop parsing input data (i.e. break).
+                            break;
+                        }
+                        index = 0;
+                        nstacks = 0;
+                    }
+                    continue
+                }
+            }
+
+            // We've run out of input data to send to the worker threads; so tell them.
+            for _ in &handles {
+                if tx_input.send(None).is_err() {
+                    // If we're unable to send data on the input channel, it means one of the
+                    // worker threads has errored and signaled to all the other worker threads
+                    // to shutdown (and they all have), in which case, as above, we should stop
+                    // trying to send information to them and proceed directly to the checking
+                    // the error channel for the error that will be there.
+                    break;
+                }
+            }
+
+            // The main thread needs to drop it's handle to the error sender here because we
+            // are about to poll the error receiver for errors, which will block until all
+            // the error senders have been dropped (including ours).
+            drop(tx_error);
+
+            // Now we poll the error channel, which will block until either all work has been
+            // completely successfully, in which case `maybe_error` will be `None` or an error
+            // has occurred on one of the worker theads, in which case `maybe_error` will be
+            // `Some(<io::Error>)`.
+            let maybe_error = rx_error.iter().next();
```

Comment:
With https://github.com/rust-lang/rust/pull/62624 we may be able to clean this up to:
```rust
rx_error.iter().next().err_or(())?;
```
one day :sweat_smile:

Any particular reason you chose to do this receive _before_ joining with `handles`?

Reply:
Question for you (as my understanding of this is probably less robust than yours): Do we even need to join the handles at all? Do we even need to drop them? Aren't we sure that by the end of `rx_error.iter().next()` that all the threads will actually exit? And isn't that the only reason why we'd join the handles at all? To ensure all the threads have exited before the main thread exits this method? I just included the loop where we're joining the handles because I was afraid there's some reason that I don't know about why that might be necessary.

Reply:
You don't _need_ to join with spawned threads, but it is handy for two reasons:

 - The return value of the threads is otherwise leaked, which causes memory to increase (slightly) over time
 - It works sort of like an assertion in that tests and such will hang if the threads for some reason _don't_ exit!

Reply:
You don't _need_ to join with spawned threads, but it is handy for two reasons:

 - The return value of the threads is otherwise leaked, which causes memory to increase (slightly) over time
 - It works sort of like an assertion in that tests and such will hang if the threads for some reason _don't_ exit!

Reply:
Awesome. Thanks for the explanation!